alf/algorithms/dqn_algorithm.py

```diff
@@ -0,0 +1,252 @@
+            use_entropy_reward=False,
+            num_critic_replicas=num_critic_replicas,
+            env=env,
+            config=config,
+            critic_loss_ctor=critic_loss_ctor,
+            # Allow custom optimizer for q_network:
+            critic_optimizer=q_optimizer,
+            alpha_optimizer=alpha_optimizer,
+            debug_summaries=debug_summaries,
+            name=name)
+        assert self._act_type == ActionType.Discrete
+
+    # Copied and modified from sac_algorithm (discrete actions).
+    def _predict_action(self,
+                        observation,
+                        state: DqnActionState,
+                        epsilon_greedy=None,
+                        eps_greedy_sampling=False,
+                        rollout=False):
+        """The reason why we want to do action sampling inside this function
+        instead of outside is that for the mixed case, once a continuous action
+        is sampled here, we should pair it with the discrete action sampled from
+        the Q value. If we just return two distributions and sample outside, then
+        the actions will not match.
+        """
+        new_state = DqnActionState()
+
+        critic_network_inputs = (observation, None)
+        if self._act_type == ActionType.Mixed:
+            critic_network_inputs = (observation, (None, continuous_action))
+
+        q_values, critic_state = self._compute_critics(
+            self._critic_networks, *critic_network_inputs, state.critic)
+
+        new_state = new_state._replace(critic=critic_state)
+        if self._act_type == ActionType.Discrete:
+            alpha = torch.exp(self._log_alpha).detach()
+        else:
+            alpha = torch.exp(self._log_alpha[0]).detach()
+        # p(a|s) = exp(Q(s,a)/alpha) / Z;
+        logits = q_values / alpha
+        discrete_action_dist = td.Categorical(logits=logits)
+        if eps_greedy_sampling:
+            discrete_action = dist_utils.epsilon_greedy_sample(
+                discrete_action_dist,
+                epsilon_greedy,
+                # NOTE: This is the only departure from SAC:
+                True)
+        else:
+            discrete_action = dist_utils.sample_action_distribution(
+                discrete_action_dist)
+
+        action_dist = discrete_action_dist
+        action = discrete_action
+
+        if (self._reproduce_locomotion and rollout
+                and not self._training_started):
+            # This uniform sampling seems important because for a squashed Gaussian,
+            # even with a large scale, a random policy is not nearly uniform.
+            action = alf.nest.map_structure(
+                lambda spec: spec.sample(outer_dims=observation.shape[:1]),
+                self._action_spec)
+
+        return action_dist, action, q_values, new_state
+
+    # Copied and modified from sac_algorithm (discrete actions).
+    def rollout_step(self, inputs: TimeStep, state: DqnState):
```

Comment:
I could do as you said, but I'll have to ignore the input epsilon_greedy to _predict_action(), because sac's rollout uses epsilon_greedy=1.0.  This makes the code harder to maintain (error prone) in the future, so I'd prefer to copy the rollout_step() function over and explicitly pass in the epsilon_greedy parameter.